burn-core/src/optim/adamw.rs

```diff
@@ -0,0 +1,369 @@
+
+            // Update second moment estimate.
+            let factor = 1.0 - self.beta_2;
+            state.moment_2 = state
+                .moment_2
+                .mul_scalar(self.beta_2)
+                .add(grad.powf(2.0).mul_scalar(factor));
+
+            // Update time.
+            state.time += 1;
+
+            state
+        } else {
+            // Initialize first moment estimate.
+            let factor = 1.0 - self.beta_1;
+            let moment_1 = grad.clone().mul_scalar(factor);
+
+            // Initialize second moment estimate.
+            let factor = 1.0 - self.beta_2;
+            let moment_2 = grad.powf(2.0).mul_scalar(factor);
+
+            AdaptiveMomentumWState::new(0, moment_1, moment_2)
+        };
+
+        let time: i32 = (state.time as i32).elem();
+
+        // Compute bias-corrected first and second moment estimates.
+        let moment_1_corrected = state
+            .moment_1
+            .clone()
+            .div_scalar(1f32 - self.beta_1.powi(time));
+
+        let moment_2_corrected = state
+            .moment_2
+            .clone()
+            .div_scalar(1f32 - self.beta_2.powi(time));
+
+        // Compute update delta. This still needs to be scaled by the learning rate.
+        let update_delta = moment_1_corrected
+            .clone()
+            .div(moment_2_corrected.clone().sqrt().add_scalar(self.epsilon));
+
+        (
+            update_delta,
+            AdaptiveMomentumWState::new(state.time, state.moment_1, state.moment_2),
+        )
+    }
+}
+
+impl<B: Backend, const D: usize> AdaptiveMomentumWState<B, D> {
+    /// Move state to device.
+    ///
+    /// # Arguments
+    ///
+    /// * `device` - Device to move state to.
+    ///
+    /// # Returns
+    ///
+    /// Returns state moved to device.
+    pub fn to_device(mut self, device: &B::Device) -> Self {
+        self.moment_1 = self.moment_1.to_device(device);
+        self.moment_2 = self.moment_2.to_device(device);
+        self
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::module::{Module, Param};
+    use crate::optim::{GradientsParams, Optimizer};
+    use crate::record::{BinFileRecorder, FullPrecisionSettings, Recorder};
+    use crate::tensor::{Data, Distribution, Tensor};
+    use crate::{nn, TestADBackend, TestBackend};
+
+    const LEARNING_RATE: LearningRate = 0.01;
+
+    #[test]
+    fn test_adamw_optimizer_save_load_state() {
+        let linear = nn::LinearConfig::new(6, 6).init();
+        let x = Tensor::<TestADBackend, 2>::random([2, 6], Distribution::Default);
+        let mut optimizer = create_adamw();
+        let grads = linear.forward(x).backward();
+        let grads = GradientsParams::from_grads(grads, &linear);
+        let _linear = optimizer.step(LEARNING_RATE, linear, grads);
+        BinFileRecorder::<FullPrecisionSettings>::default()
+            .record(optimizer.to_record(), "/tmp/test_optim".into())
```

Comment:
I think it's a good opportunity to improve the usage of temporary directories. Could we please use the `TempDir` from the `tempfile` crate? I realize this is how the previous tests were done, and I wish to refactor them later.

Here's what you'll need to do:
Add the following to burn-core's `Cargo.toml`:
```
[dev-dependencies]
tempfile = {workspace = true}
```
You can follow this usage example from [TempDir::new()](https://docs.rs/tempfile/latest/tempfile/struct.TempDir.html#method.new):

```rust
use tempfile::TempDir;
let temp_dir = TempDir::new().unwrap();
```

The advantage of this approach is that it is cross-platform, and temporary directories are removed after the `temp_dir` variable goes out of scope.